scipy/stats/_discrepancy.pyx

```diff
@@ -0,0 +1,502 @@
+    >>> disc_init
+    0.04769081147119336
+    >>> qmc.update_discrepancy(space[-1], space[:-1], disc_init)
+    0.008142039609053513
+
+    """
+    sample = np.asarray(sample)
+    x_new = np.asarray(x_new)
+
+    # Checking that sample is within the hypercube and 2D
+    if not sample.ndim == 2:
+        raise ValueError('Sample is not a 2D array')
+
+    if not (np.all(sample >= 0) and np.all(sample <= 1)):
+        raise ValueError('Sample is not in unit hypercube')
+
+    # Checking that x_new is within the hypercube and 1D
+    if not x_new.ndim == 1:
+        raise ValueError('x_new is not a 1D array')
+
+    if not (np.all(x_new >= 0) and np.all(x_new <= 1)):
+        raise ValueError('x_new is not in unit hypercube')
+
+    if x_new.shape[0] != sample.shape[1]:
+        raise ValueError('x_new and Sample must be broadcastable')
+
+    return c_update_discrepancy(x_new, sample, initial_disc)
+
+
+cdef double c_update_discrepancy(double[::1] x_new_view,
+                                 double[:, ::1] sample_view,
+                                 double initial_disc):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0] + 1
+        Py_ssize_t xnew_nlines = x_new_view.shape[0]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double prod = 1, tmp_sum= 0
+        double  disc1 = 0, disc2 = 0, disc3 = 0
+        double[::1] abs_ = np.zeros(n, dtype=np.float64)
+
+
+    # derivation from P.T. Roy (@tupui)
+    for i in range(xnew_nlines):
+        abs_[i] = fabs(x_new_view[i] - 0.5)
+        prod *= (
+                    1 + 0.5 * abs_[i]
+                    - 0.5 * pow(abs_[i], 2)
+        )
+
+    disc1 = (- 2 / <double> n) * prod
+
+    prod = 1
+    for i in range(n - 1):
+        for j in range(xnew_nlines):
+            prod *= (
+                1 + 0.5 * abs_[j] +
+                0.5 * fabs(sample_view[i, j] - 0.5) -
+                0.5 * fabs(x_new_view[j] - sample_view[i, j])
+            )
+        disc2 += prod
+        prod = 1
+
+    disc2 *= 2 / pow(n, 2)
+
+    for i in range(xnew_nlines):
+        prod *= 1 + abs_[i]
+
+    disc3 = 1 / pow(n, 2) * prod
+
+    return initial_disc + disc1 + disc2 + disc3
+
+
+ctypedef double (*func_type)(double[:, ::1], Py_ssize_t,
+                             Py_ssize_t) nogil
+
+
+cdef double threaded_loops(func_type loop_func, double[:,
+                         ::1] sample_view, unsigned int workers):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        double disc2 = 0
+
+    if workers <= 1:
+        return loop_func(sample_view, 0, n)
+
+    cdef:
+        vector[thread] threads
+        unsigned int tid
+        Py_ssize_t istart, istop
+
+    for tid in range(workers):
+        istart = <Py_ssize_t> (n / workers * tid)
+        istop = <Py_ssize_t> (
+                n / workers * (tid + 1)) if tid < workers - 1 else n
+        threads.push_back(
+            thread(one_thread_loop, loop_func, ref((&disc2)[0]),
```

Comment:
Do you really need `ref((&disc2)[0])`? I think `ref(disc2)` is enough. It seems to be a workaround to a bug, do you have a reference about this cython bug?

Reply:
Changes in 90d7d72c4722fbff735d46e01b218d45693086aa